### ISSUE TITLE
Proposal: Use List Semantics in GetLeavesByIndex

### ISSUE THREAD

**Issue by gdbelvin** [google/trillian#448]
The [Google API style guide](https://cloud.google.com/apis/design/standard_methods) suggest using a start and length to enumerate a list. Updating the Trillian API slightly to support this would allow the Trillian API to be easily enumerated over HTTP and allow simplified error processing.

Today
`GetLeavesByIndex` takes a list of indexes to query and returns a list of leaves, each of which can individually fail.  

`GetLeavesByIndex([]index int64) ([]*Leaf, error)` 

Proposal
Modify `GetLeavesByIndex` to take a start index and a count, and return an optionally shorter list of leaves, and only support failures for the query as a whole. 

`GetLeavesByIndex(startIndex, count int64) ([]*Leaf, error)`

Rational
- The list API design is a closer match to the 6962 `getEntries(start, end)` semantics. 
- Simplifies processing the response since the responses are defined to be in-order. 
- Simplifies error processing by only supporting an error for the whole batch. 
   - Alternatively, an error occurring midway through processing could just return the valid items retrieved thus far. Returning fewer items than requested is explicitly supported by the API. 
- Makes the API easier to expose over standard HTTP methods. 

Would love your thoughts.

**Comment by Martin2112** [comment]
Returning fewer items than requested is allowed but we don't want to hide errors by returning partial batches if there is an underlying reason why the the remaining ones can't be retrieved. 

One of the logs that failed was returning an invalid incomplete set of results with missing items from the range.

**Comment by gdbelvin** [comment]
That's an interesting note about how logs can fail.  It sounds like the returned set was incomplete?  This API definition requires that all returned sets be contiguous so there's no confusion. 

Happy to either way on whether to fail a whole range, or to return the first valid section. 
A client will be returned the error when it comes back to ask for the next page of data.

**Comment by Martin2112** [comment]
There's a distinction between what an API requires and what actually happens when bugs or storage failures occur in practice.

Not every client is a monitor that's scanning the whole log. There may not be a next request that exposes the error. We want failures to be obvious so it should not return partial data if there's an error.

Also this removes the ability to request non contiguous leaves in a single request. That doesn't matter for 6962 but is again constraining what the API supports.

**Comment by gdbelvin** [comment]
Agreeing that servers should return errors for the requests that trigger
them, rather than truncating results.

**Comment by codingllama** [comment]
First of all, let me start with API style nitpicking: *"(API style) (...) suggest using a start and length to enumerate a list"*. Actually, the standard filters for [list methods](https://cloud.google.com/apis/design/standard_methods#list) are `page_size` and `page_token`. The page size is not a total length, but the max results in a single page, and the page token is opaque and only set after the first request, so it doesn't work as a start index.

What you're talking about is a [custom method](https://cloud.google.com/apis/design/custom_methods), not a standard list.

In terms of API flexibility, as @Martin2112 suggested, `GetLeavesByIndex([]int64)` is more flexible than `GetLeavesByIndex(start, length int64)`. What strikes me as a bit odd is that we don't have single-leaf get methods, which would fix some of the arguments in this discussion.

Putting it all together, I'd suggest an API that resembles this:

```proto
service TrillianLog {
  // Single-entity get methods ("standard")
  rpc GetLeafByIndex (GetLeafsByIndexRequest) returns (Leaf) {}
  rpc GetLeafByHash (GetLeafByHashRequest) returns (Leaf) {}

  // Batch-read methods (custom)
  rpc GetLeavesByIndex (GetLeavesByIndexRequest) returns (GetLeavesByIndexResponse) {}
  rpc GetLeavesByHash (GetLeavesByHashRequest) returns (GetLeavesByHashResponse) {}
  rpc GetLeavesByRange (GetLeavesByRangeRequest) returns (GetLeavesByRangeResponse) {}
}

message GetLeavesByRangeRequest {
  int64 tree_id = 1;
  int64 start_index = 2;
  int64 count = 3;
}
```

Where `GetLeavesByRange` is what @gdbelvin suggested.

The question that remains is: do we need *all* of those methods? Both of you have better insights into CT / Trillian clients than me. I'd say that, from an API design point-of-view, we should definitely supply single-leaf get methods and only add the batching we *really* need. For example, which needs are we trying to address with those methods? Do we expect personalities to read leaves in batch? Do we expect monitors to work primarily with ranges? Is there something else we're trying to address?

I've had before APIs that grew wild in functionality by adding more RPCs and more behaviors to existing RPCs. Very few of those paid their cost in practice. I'd suggest we push for an API that's simple but complete, ie, that offers the necessary primitive building blocks for clients without absorbing unneeded complexity.

**Comment by Martin2112** [comment]
Was "GetLeafsByIndexRequest" a typo or intentional?

CT needs a single leaf for get-entry-and-proof. It's a debug API really but
it's in the RFC so has to be implemented.

A batch read is needed to efficiently implement the get-entries request.
There are several applications that want to scan a log from some starting
point to the end.

Martin

**Comment by AlCutter** [comment]
By way of an example of a use-case for why we have `ByIndex` consider the Map example under `examples/ct`; this is a map of log indices by domain, you could imagine an app built around this to need to fetch the certs stored at the indices when a user/app queries for a given domain, and it'd make sense to batch that request.

Other things to think about from the platform API design guide are:
 - the List pattern has no concept of a start index, just `orderBy`, `pagination`, and `filter`
 - see the comments on range specification here: https://cloud.google.com/apis/design/design_patterns (use half-open, @pphaneuf is going to love that.)
 - would `filter` be the platform _idiomatic_ approach for this (odd as it feels compared to, say, 6962)?

Also `/Leafs/Leaves/` :)

**Comment by codingllama** [comment]
Leafs was a typo! :)

Re pagination and page_token: it was just a nit. No need to add pagination if we don't need it. (Note that pagination is not the same as a start/end index.)

On list methods and ListLeavesByIndex: as @AlCutter hinted, it's not exactly an standard list method. (More API nitpicking follows) The standard list method would simply be ListLeaves and likely use the [filter pattern](https://cloud.google.com/apis/design/naming_convention#list_filter_field). We could go that route and have a filter that says "index > x", but a custom batch read tends to be simpler to support.

(Thanks for the [representing ranges](https://cloud.google.com/apis/design/design_patterns#representing_ranges) link.)

So, summing up, we want:
* get-by-index (KT)
* batch-list-by-range (CT, KT, monitors, get-entries)
* batch-list-by-indexes (examples/ct map)

Did I miss anything? In particular, what about the "*ByHash" methods?

**Comment by Martin2112** [comment]
Is this work ready to be done? And do we think it should be done sooner rather than later?

**Comment by Martin2112** [comment]
I think this is still a nice to have and we also have added a get by range API that might reduce the need for it? I will close for now and we can reopen later.
